E0623 gives very confusing message

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following code: <!-- Please provide a link to play.rust-lang.org -->

```rust
fn bar(writer: &mut Writer) {
    baz(writer.indent());
    writer.write("world");
}

fn baz(writer: &mut Writer) {
    writer.write("hello");
}

pub struct Writer<'a> {
    target: &'a mut String,
    indent: usize,
}

impl<'a> Writer<'a> {
    fn indent(&'a mut self) -> &'a mut Self {
        &mut Self {
            target: self.target,
            indent: self.indent + 1,
        }
    }

    fn write(&mut self, s: &str) {
        for _ in 0..self.indent {
            self.target.push(' ');
        }
        self.target.push_str(s);
        self.target.push('\n');
    }
}
```

The current output is:

```
error[E0623]: lifetime mismatch
 --> src/writer.rs:2:16
  |
1 | fn bar(writer: &mut Writer) {
  |                -----------
  |                |
  |                these two types are declared with different lifetimes...
2 |     baz(writer.indent());
  |                ^^^^^^ ...but data from `writer` flows into `writer` here
```

This is really confusing. The error mentions two types, but then it only points to one type. And then it says "data from `writer` flows into `writer`". What? The data flows into itself?

Minimal repro ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7f3fc8b17f3d058212486a8cc76c5d28)):

```rust
struct Foo<'a>(&'a i32);

impl<'a> Foo<'a> {
    fn modify(&'a mut self) {}
}

fn bar(foo: &mut Foo) {
    foo.modify();
}
```

```
error[E0623]: lifetime mismatch
 --> src/lib.rs:8:9
  |
7 | fn bar(foo: &mut Foo) {
  |             --------
  |             |
  |             these two types are declared with different lifetimes...
8 |     foo.modify();
  |         ^^^^^^ ...but data from `foo` flows into `foo` here
```

* Getting rid of the `'a` on `&'a mut self` makes it compile.
* Getting rid of the `mut` on `&'a mut self` makes it compile.
* Getting rid of the lifetime parameter on `Foo` makes it compile.

If we put explicit lifetime annotations in `bar`'s function signature, we still get the same error:

```rust
fn bar<'a, 'b>(foo: &'a mut Foo<'b>) {
    foo.modify();
}
```

```
error[E0623]: lifetime mismatch
 --> src/lib.rs:8:9
  |
7 | fn bar<'a, 'b>(foo: &'a mut Foo<'b>) {
  |                     ---------------
  |                     |
  |                     these two types are declared with different lifetimes...
8 |     foo.modify();
  |         ^^^^^^ ...but data from `foo` flows into `foo` here
```

However, if we annotate only `Foo`, then we get a much better error:

```rust
fn bar<'a>(foo: &mut Foo<'a>) {
    foo.modify();
}
```

```
error[E0621]: explicit lifetime required in the type of `foo`
 --> src/lib.rs:8:9
  |
7 | fn bar<'a>(foo: &mut Foo<'a>) {
  |                 ------------ help: add explicit lifetime `'a` to the type of `foo`: `&'a mut Foo<'a>`
8 |     foo.modify();
  |         ^^^^^^ lifetime `'a` required
```

And indeed, that is the true fix ([playgound](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ab404b28bda93a07e6a557aee530d966)):

```rust
fn bar<'a>(foo: &'a mut Foo<'a>) {
    foo.modify();
}
```

As I mentioned above, the error is contingent upon `Foo::modify` taking in a mutable reference. Perhaps this is due to the [invariance](https://doc.rust-lang.org/nomicon/subtyping.html#variance) of `&mut`. Very tricky.

In any case, I think we should try to find a way to display a message that is very similar to the `add explicit lifetime` message.

@rustbot label +D-confusing +D-terse +D-newcomer-roadblock +A-lifetimes

--- a/tests/ui/lifetimes/lifetime-errors/ex3-both-anon-regions-one-is-struct-5.rs
+++ b/tests/ui/lifetimes/lifetime-errors/ex3-both-anon-regions-one-is-struct-5.rs
@@ -1,0 +1,13 @@
+// Regression test for #91831
+
+struct Foo<'a>(&'a i32);
+
+impl<'a> Foo<'a> {
+    fn modify(&'a mut self) {}
+}
+
+fn bar(foo: &mut Foo) {
+    foo.modify(); //~ ERROR lifetime may not live long enough
+}
+
+fn main() {}

--- a/tests/ui/lifetimes/lifetime-errors/ex3-both-anon-regions-one-is-struct-5.stderr
+++ b/tests/ui/lifetimes/lifetime-errors/ex3-both-anon-regions-one-is-struct-5.stderr
@@ -1,0 +1,20 @@
+error: lifetime may not live long enough
+  --> $DIR/ex3-both-anon-regions-one-is-struct-5.rs:10:5
+   |
+LL | fn bar(foo: &mut Foo) {
+   |        ---  - let's call the lifetime of this reference `'1`
+   |        |
+   |        has type `&mut Foo<'2>`
+LL |     foo.modify();
+   |     ^^^^^^^^^^^^ argument requires that `'1` must outlive `'2`
+   |
+   = note: requirement occurs because of a mutable reference to `Foo<'_>`
+   = note: mutable references are invariant over their type parameter
+   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
+help: consider introducing a named lifetime parameter
+   |
+LL | fn bar<'a>(foo: &'a mut Foo<'a>) {
+   |       ++++       ++        ++++
+
+error: aborting due to 1 previous error
+

